### PR TITLE
bugfix: add routes for excluded ips

### LINKED ISSUE #30: Bad path from overlay pod to gateway of underlay subnets
Bug Report
---

Type: *bug report*

### What happened
Bad network path from overlay pod to gateway of underlay subnets, when ping underlay gateway from overlay pod, unexpected neighbor event comes from vxlan interface.

### What you expected to happen
We expect overlay pod will react underlay gateway on a underlay+snat network path.

### How to reproduce it (as minimally and precisely as possible)
Ping underlay gateway from overlay pod.

### Anything else we need to know?
Beside underlay gateway, if we only use part of a underlay subnet, the rest IPs will have the same problem as gateway.

### Environment
- rama version:
- OS (e.g. `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Kubernetes version:
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by hhyasdf**
Pull Request Description
---

### Describe what this PR does / why we need it
Rama support a flexible ip range configuration, in which even a discontinuous ip address range can be described. While if a subnet CIDR has some excluded ip addresses, there would be some problem for overlay pod to access these excluded ip addresses.

### Does this pull request fix one issue?

Fixes #30 

### Describe how you did it
Four cases should be considered:

1. Underlay pod to access underlay excluded ip addresses: the same with underlay pod' accessing other underlay addresses, not changed.
2. Underlay pod to access overlay excluded ip addresses: this traffic should never pass through vxlan interface.
3. Overlay pod to access overlay excluded ip addresses: the same with overlay pod's accessing ip addresses outside the cluster, should never pass through vxlan interface.
4. Overlay pod to access underlay excluded ip addresses: the same with overlay pod's accessing ip addresses outside the cluster, should never pass through vxlan interface.

So, in the PR, I add some `throw` routes to handle the traffic with excluded destination address blocks, to make traffic don't pass through the vxlan interface if either it's destination is a overlay excluded ip addresses or it's the traffic from overlay pod to underlay excluded ip addresses.

### Describe how to verify it
Use overlay pod to access underlay gateway or excluded ip addresses.

### Special notes for reviews

### CODE REVIEW

File: pkg/daemon/route/route.go [hunk 1]
```diff
@@ -165,33 +161,67 @@ func CreateRouteManager(localDirectTableNum, toOverlaySubnetTableNum, overlayMar
...
+		subnetInfo.reservedIPs = append(subnetInfo.reservedIPs, reservedIPs...)
+	}
+
+	if len(excludeIPs) != 0 {
+		subnetInfo.excludeIPs = append(subnetInfo.excludeIPs, excludeIPs...)
+	}
+
+	if start != nil || end != nil {
+		if start == nil {
+			start = cidr.IP
```
Comment (mars1024) [inline pkg/daemon/route/route.go:204]: actually we are using the second IP of network IP as start

File: pkg/daemon/route/route.go [hunk 2]
```diff
@@ -165,33 +161,67 @@ func CreateRouteManager(localDirectTableNum, toOverlaySubnetTableNum, overlayMar
...
+		}
+
+		if end == nil {
+			end = daemonutils.LastIP(cidr)
```
Comment (mars1024) [inline pkg/daemon/route/route.go:208]: you could use the functions as IPAM, no similar functions in same project

File: pkg/daemon/route/route.go (hunk 1 above)
Comment (hhyasdf) [reply pkg/daemon/route/route.go:204]: This is used for calculate a CIDR, not available ip to allocate, every possible ip should be considered.

File: pkg/daemon/route/route.go (hunk 2 above)
Comment (hhyasdf) [reply pkg/daemon/route/route.go:208]: For now, this part of functions is only used for parse an ip range in daemon, maybe is not necessary until we actually find it's a common usage.

File: pkg/daemon/utils/ip_range.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+func calculateIPLastZeroBits(ip net.IP) int {
+	zeroBits := 0
+	for {
+		nextPossibleZeroBits := zeroBits + 1
+		ipInt := ipToInt(ip)
+		ipUint64 := ipInt.Uint64()
+
+		if ipUint64 == (ipUint64>>nextPossibleZeroBits)<<nextPossibleZeroBits {
```
Comment (mars1024) [inline pkg/daemon/utils/ip_range.go:163]: 这里没有必要每次搞个新的 ipUint64 啊，位信息还在，一直左移一位+比较就可以了

File: pkg/daemon/utils/ip_range.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+		start: start,
+		end:   end,
+	}, nil
+}
+
+func (ir *IPRange) TryAddIP(ipAddr net.IP) (success bool) {
+	if ipAddr.Equal(ip.PrevIP(ir.start)) {
+		ir.start = ipAddr
+		return true
+	} else if ipAddr.Equal(ip.NextIP(ir.end)) {
```
Comment (mars1024) [inline pkg/daemon/utils/ip_range.go:36]: no need else here

File: pkg/daemon/utils/ip_range.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+		}
+
+		if ipRange != nil {
+			excludeIPRanges = append(excludeIPRanges, ipRange)
+		}
+	}
+
+	allExcludedIPs := excludeIPs
+	if len(reservedIPs) != 0 {
+		allExcludedIPs = append(allExcludedIPs, reservedIPs...)
```
Comment (mars1024) [inline pkg/daemon/utils/ip_range.go:108]: reservedIPs should be kept not excluded

File: pkg/daemon/utils/ip_range.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+	if start.To4() == nil {
+		ipLen = net.IPv6len * 8
+	}
+
+	minCidrPrefixLen := ipLen - zeroBits
+	maxValidCidrPrefixLen := minCidrPrefixLen
+
+	for {
+		tmpCidr := &net.IPNet{
+			IP:   start,
```
Comment (mars1024) [inline pkg/daemon/utils/ip_range.go:186]: should IP be masked by Mask to be a standard IPNet?

File: pkg/daemon/route/utils.go [hunk 7]
```diff
@@ -295,32 +302,106 @@ func ensureFromPodSubnetRuleAndRoutes(forwardNodeIfName string, cidr *net.IPNet,
...
+		localAddrList, err := netlink.AddrList(nil, family)
+		if err != nil {
+			return fmt.Errorf("list local addresses failed: %v", err)
+		}
+
+		isLocalSubnet := false
+		for _, address := range localAddrList {
+			if cidr.Contains(address.IP) {
+				// Check if address is an enhanced address or used to connect a subnet.
+				if address.Flags&unix.IFA_F_NOPREFIXROUTE == 0 {
```
Comment (mars1024) [inline pkg/daemon/route/utils.go:322]: if this judgment locate the local address except enhanced addresses? why?

File: pkg/daemon/utils/ip_range.go (hunk 5 above)
Comment (hhyasdf) [reply pkg/daemon/utils/ip_range.go:108]: ReservedIPs can not be allocated to pod either, and also can be used to "exclude" ip, so I think is better to keep it.

File: pkg/daemon/route/utils.go (hunk 7 above)
Comment (hhyasdf) [reply pkg/daemon/route/utils.go:322]: Enhanced addresses will always have the `IFA_F_NOPREFIXROUTE` flag.

File: pkg/daemon/utils/ip_range.go (hunk 6 above)
Comment (hhyasdf) [reply pkg/daemon/utils/ip_range.go:186]: No need to do this, the last `zeroBits` must be zero, we don't need to mask it again.

File: pkg/daemon/utils/ip_range.go (hunk 4 above)
Comment (hhyasdf) [reply pkg/daemon/utils/ip_range.go]: fixed

File: pkg/daemon/utils/ip_range.go (hunk 5 above)
Comment (hhyasdf) [reply pkg/daemon/utils/ip_range.go]: fixed

File: pkg/daemon/utils/ip_range.go (hunk 6 above)
Comment (mars1024) [reply pkg/daemon/utils/ip_range.go:187]: Mask is changing every loop, so we could get some nonstandard CIDR like 192.168.0.4/24, 192.168.0.4/25, 192.168.0.4/26 ...

File: pkg/daemon/utils/ip_range.go [hunk 8]
```diff
@@ -0,0 +1,227 @@
+package utils
```
Comment (mars1024) [inline pkg/daemon/utils/ip_range.go:1]: missing copyright here

File: pkg/daemon/containernetwork/utils.go [hunk 9]
```diff
@@ -235,3 +235,55 @@ func ListLocalAddressExceptLink(exceptLinkName string) ([]netlink.Addr, error) {
...
+	for _, neigh := range neighList {
+		if neigh.IP.Equal(podIP) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func checkPodNetConfigReady(podIP net.IP, podCidr *net.IPNet, forwardNodeIfIndex int, family int) error {
```
Comment (mars1024) [inline pkg/daemon/containernetwork/utils.go:269]: should we retry some times here to wait an async ready? I don't think it is good to let kubelet throw out the unnecessary errors.

File: pkg/daemon/utils/ip_range.go (hunk 6 above)
Comment (mars1024) [reply pkg/daemon/utils/ip_range.go:187]: mistake, gotcha your point

File: pkg/daemon/containernetwork/utils.go (hunk 9 above)
Comment (hhyasdf) [reply pkg/daemon/containernetwork/utils.go:270]: fixed

File: pkg/daemon/utils/ip_range.go (hunk 8 above)
Comment (hhyasdf) [reply pkg/daemon/utils/ip_range.go:17]: fixed

**Review by mars1024 — APPROVED** [review]
/lgtm
/approve
Thanks!

**Comment by mars1024** [comment]
ref to #30

**Pull request merged**